### ISSUE TITLE
Ability to get Datatable without using Http\Requests

### ISSUE THREAD

**Issue by najela** [yajra/laravel-datatables#223]
I am currently writing a laravel Job that retrives the datatable and Sends the JSON data to a webhook.

 The problem is this datatable package uses the `Http\Requests`  which is not being populated when my cron Job is running. I could always pass on this data to my cron Job but it would not available in the Laravel Request object.

Is it possible to supply the request parameters through another means?

**Comment by yajra** [comment]
You would need to make a request object. Try something like.

```php
use yajra\Datatables\Request;
........

$request = app(Request::class);
$request->merge(['length' => -1]); // populate data

```

**Comment by najela** [comment]
@yajra  Thanks very much for Replying.

The thing is I don't know where to put this code. For instance, This is the `fire` method in my Laravel Job

```php
public function fire($job, array $data)
{
    $user = User::all();
    $hello = Datatable::of($user);
 
    return $hello;
}
```

I noticed that the Request class is being injected dynamically in the Datatable class. So I was wondering where I could do this modification

**Comment by yajra** [comment]
@najela, what version of Laravel are you using? On Laravel 5 you can do something like below:

```php
namespace App\Jobs;

use App\Jobs\Job;
use App\Models\User;
use Illuminate\Contracts\Bus\SelfHandling;
use yajra\Datatables\Datatables;

class DemoJob extends Job implements SelfHandling
{
    /**
     * @var \yajra\Datatables\Datatables
     */
    private $datatables;

    /**
     * Create a new job instance.
     *
     * @param \yajra\Datatables\Datatables $datatables
     */
    public function __construct(Datatables $datatables)
    {
        $this->datatables = $datatables;
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        $request = $this->datatables->getRequest();
        $request->merge(['length' => -1]);

        $user = User::all();
        return $this->datatables->of($user);
    }
}
```

**Comment by najela** [comment]
Sorry for the late reply. I tried the above solution and It didn't work for me.

Here is my wrapper for the `yajra` datatable

```php
<?php

namespace Passport\CustomerMeta\Datatable;

use ClientsModel;
use Passport\Meta\Contracts\ObjectMetaRepository;
use yajra\Datatables\Datatables;
use LOG;

class YajraCustomerMetaDatatable implements CustomerMetaDatatable
{
    /**
     * @var Array
     */
    private $accessibleClients;

    /**
     * @var ObjectMetaRepository
     */
    private $meta;

    /**
     * var ClientsModel.
     */
    private $clients;

    /**
     * @var \yajra\Datatables\Datatables
     */
    private $datatables;

    /**
     * Set dependencies.
     *
     * @param ObjectMetaRepository $objectMetaRepository
     * @param ClientsModel         $clients
     */
    public function __construct(
        ObjectMetaRepository $objectMetaRepository,
        ClientsModel $clientsModel,
        Datatables $datatables
    ) {
        $this->clients = $clientsModel;
        $this->meta = $objectMetaRepository;
        $this->datatables = $datatables;
    }

    /**
     * Returns the Datatabe JSON array.
     *
     * @param $user
     * @param $partyId
     * @param $clientId
     *
     * @return string
     */
    public function make($user, $partyId, $clientId, $input = null)
    {
        $request = $this->datatables->getRequest();
        if (($user->isManager())) {
            $clientTreeInfo = $this->clients->fetchFlatTree($clientId, 1);
        } else {
            $clientTreeInfo = $this->clients->fetchFlatTreeByUser($user->getId(), 7, $clientId);
        }
        $clientIds[] = $clientId;
        if ($clientTreeInfo) {
            foreach ($clientTreeInfo as $client) {
                $clientIds[] = $client['client_id'];
            }
        }

        if (!is_null($input)) {
            $request->merge($input);
        }

        $meta = $this->meta->datatable($clientIds, $partyId);

        $hello = $this->datatables->of($meta)->make();

        if (is_null($input)) {
            file_put_contents('/private/var/www/work/maatwebsite/passport/logs/temp/good.log', var_export($hello, true));
            file_put_contents('/private/var/www/work/maatwebsite/passport/logs/temp/good-request.log', var_export($request->all(), true));
        } else {
            file_put_contents('/private/var/www/work/maatwebsite/passport/logs/temp/bad.log', var_export($hello, true));
            file_put_contents('/private/var/www/work/maatwebsite/passport/logs/temp/bad-request.log', var_export($request->all(), true));
        }

        return $hello;
    }
}
```

Now when I call the `make()` method of the above class froma a controller with the request passed automatically, I get the correct output

```php
return $this->datatable->make($USER, $partyId, $clientId);
```

But when I call this same method from my cron, I get a different output

```
$data = $this->datatable->make($USER, $data['partyId'], $data['clientId'], $data['input']);
```

If I do an `ls` in the directory containing the logs I made for debugging, I see that the Request variable in both cases is the same. What differs is the datatable that is generated.


```bash
[najela@Achos-MacBook-Pro:temp]$ ls -lh
total 1.7M
-rw-r--r-- 1 najela staff 5.1K Oct 24 08:28 bad-request.log
-rw-r--r-- 1 najela staff 1.7M Oct 24 08:28 bad.log
-rw-r--r-- 1 najela staff 5.1K Oct 24 08:28 good-request.log
-rw-r--r-- 1 najela staff  11K Oct 24 08:28 good.log
```

You can see a significant difference  in the size of the outputs.

Thanks oncemore for your help

**Comment by najela** [comment]
I am using `v5.11.10` of yajra datatable

**Comment by yajra** [comment]
@najela, I'm a bit confused. v5.11.10 of datatable is for Laravel 5 and your Job is Laravel 4? I think that is the reason why the job fails? Laravel 4 datatable version does not have `Request` object. It's only available on L5 versions.

**Comment by najela** [comment]
@yajra  Its not a Laravel Project. Its a personal project which uses laravel components. I use both the laravel 4 queue and the Laravel 5 queue. I'm upgrading the project to laravel 5 but Its too large so things can't be done that fast thats why I maintain some laravel 4 components. I could implement the queue in laravel 5 but I really don't think that would solve the problem. 

Any other tips?

**Comment by yajra** [comment]
Wow, I haven't personally used the package outside Laravel....

Anyways, I just remember that some weird bug on `Request` class of Laravel 4 when I was writing a test. The request variables are not being parsed properly when executed in the console and I think was fixed on Laravel 5. This maybe on your case too?

Lastly, another snippet that I used in unit testing. Maybe try initializing your `Datatables` like below?
```php
$datatables = new Datatables(Request::capture());
```
